src/main/scala/com/fulcrumgenomics/alignment/NeedlemanWunschAligner.scala

```diff
@@ -0,0 +1,187 @@
+    */
+  protected def buildMatrices(query: Array[Byte], target: Array[Byte]): (Matrix[Int], Matrix[Direction]) = {
+    val scoring = Matrix[Int](query.length + 1, target.length + 1)
+    val trace   = Matrix[Direction](query.length + 1, target.length + 1)
+
+    // Top left corner
+    scoring(0,0) = 0
+    trace(0,0)   = Done
+
+    // Down the left hand side
+    forloop(from=1, until=query.length+1) { i =>
+      trace(i, 0)   = Up
+      scoring(i, 0) = scoring(i-1, 0) + gapExtend + (if (i == 1) gapOpen else 0)
+    }
+
+    // Along the top
+    forloop(from=1, until=target.length+1) { j =>
+      trace(0, j)   = Left
+      scoring(0, j) = scoring(0, j-1) + gapExtend + (if (j == 1) gapOpen else 0)
+    }
+
+    // The interior of the matrix
+    forloop(from=1, until=query.length+1) { i =>
+      forloop(from=1, until=target.length+1) { j =>
+        val diagonalScore = scoring(i-1, j-1) + scoringFunction(query(i-1), target(j-1))
+        val upScore       = scoring(i-1, j  ) + gapExtend + (if (trace(i-1, j  ) == Up) 0 else gapOpen)
+        val leftScore     = scoring(i,   j-1) + gapExtend + (if (trace(i,   j-1) == Left) 0 else gapOpen)
+
+        val maxScore = max(max(diagonalScore, upScore), leftScore)
+        scoring(i,j) = maxScore
+        trace(i,j)   = if (maxScore == diagonalScore) Diagonal else if (maxScore == leftScore) Left else Up
+      }
+    }
+
+    (scoring, trace)
+  }
+
+  /**
+    * Given the scoring and trace back matrices, construct the [[Alignment]] object for the (or one of the)
+    * optimal alignments.
+    *
+    * @param query the query sequence
+    * @param target the target sequence
+    * @param scoring the scoring matrix
+    * @param trace the trace back matrix
+    * @return an [[Alignment]] object representing the alignment
+    */
+  protected def generateAlignment(query: Array[Byte], target: Array[Byte], scoring: Matrix[Int], trace: Matrix[Direction]): Alignment = {
+    var i = query.length
+    var j = target.length
+    var currOperator: CigarOperator = null
+    var currLength: Int = 0
+    val stack = new mutable.Stack[CigarElem]()
+
+    while (i != 0 || j != 0) {
+      val op = trace(i,j) match {
+        case Done     =>
+          i = 0
+          j = 0
+          null
+        case Up       =>
+          i -= 1
+          CigarOperator.INSERTION
+        case Left     =>
+          j -= 1
+          CigarOperator.DELETION
+        case Diagonal =>
+          i -= 1
+          j -= 1
+          if (query(i) == target(j)) this.matchOp else this.mismatchOp
+      }
+
+      val extending = op == currOperator
+
+      if (extending) {
+        currLength += 1
+        if (i == 0 && j == 0) stack.push(CigarElem(currOperator, currLength))
+      }
+      else {
+        if (currLength > 0) stack.push(CigarElem(currOperator, currLength))
+        currOperator = op
+        currLength   = 1
+        if (i == 0 && j == 0) stack.push(CigarElem(currOperator, currLength))
```

Comment:
Can move this to after the `if/else` and remove the same call within the `if/else`:
```
if (i == 0 && j == 0) stack.push(CigarElem(currOperator, currLength))
```